qiita_db/util.py

```diff
@@ -2311,3 +2317,337 @@ def send_email(to, subject, body):
+        timedelta(seconds=float(min(y_plot)))
+
+    x_plot = np.array(_df[COL_NAME])
+    failures_df = _resource_allocation_failures(_df, k, a, b, best_model, curr)
+    failures = failures_df.shape[0]
+
+    ax.scatter(failures_df[COL_NAME], failures_df[curr], color='red', s=3,
+               label="failures")
+
+    ax.set_title(f'{cname}: {sname}\n real: {mini} || {maxi}\n'
+                 f'calculated: {cmin} || {cmax}\n'
+                 f'failures: {failures}')
+    ax.legend(loc='upper left')
+    return best_model, options
+
+
+def _resource_allocation_calculate(_df, x, y, models, type_, depth):
+    """Helper function for resource allocation plot. Calculates best_model and
+    best_result given the models list and x,y data.
+
+    Parameters
+    ----------
+    x: pandas.Series (pandas column), required
+        Represents x data for the function calculation
+    y: pandas.Series (pandas column), required
+        Represents y data for the function calculation
+    models: list, required
+        List of functions that will be used for visualization
+    depth: int, required
+        Maximum number of iterations in binary search
+
+    Returns
+    ----------
+    best_model: function
+        best fitting function for the current list models
+    best_result: object
+        object containing constants for the best model (e.g. k, a, b in kx+b*a)
+    """
+
+    init = [1, 1, 1]
+    best_model = None
+    best_result = None
+    best_failures = np.inf
+    best_max = np.inf
+    counter = 0
+    for model in models:
+        counter += 1
+        sl = 2000
+        sr = 100000
+        left = sl
+        right = sr
+        prev_failures = np.inf
+        min_max = np.inf
+        cnt = 0
+
+        res = [1, 1, 1]  # k, a, b
+
+        while left < right and cnt < depth:
+            middle = (left + right) // 2
+            options = minimize(_resource_allocation_custom_loss, init,
+                               args=(x, y, model, middle))
+            k, a, b = options.x
+            failures_df = _resource_allocation_failures(_df, k, a, b, model,
+                                                        type_)
+            y_plot = model(x, k, a, b)
+            cmax = max(y_plot)
+            cmin = min(y_plot)
+            failures = failures_df.shape[0]
+
+            if failures < prev_failures:
+                prev_failures = failures
+                right = middle
+                min_max = cmax
+                res = options
+
+            elif failures > prev_failures:
+                left = middle
+            else:
+                if cmin < 0:
+                    left = middle
+                elif cmax < min_max:
+                    min_max = cmax
+                    res = options
+                    right = middle
+                else:
+                    right = middle
+
+            if left >= right and cnt < depth:
+                sl += 10000
+                sr += 10000
```

Comment:
could you add a small comment explaining the increments?

Reply:
I added these because I don't have a formal proof why binary search would always result in the optimal constant, hence, I do the search in several windows up searching for the best constant until i reach the total number of iterations.